packages/glimmer-runtime/lib/compiled/opcodes/content.ts

```diff
@@ -27,12 +39,20 @@ export class AppendOpcode extends Opcode {
   evaluate(vm: VM) {
     let reference = vm.frame.getOperand();
 
-    let mapped = map(reference, normalizeTextValue);
+    let mapped = map(reference, normalizeTextOrTrustedValue);
     let cache = new ReferenceCache(mapped);
-    let node = vm.stack().appendText(cache.peek());
-
-    if (!isConst(reference)) {
-      vm.updateWith(new UpdateAppendOpcode(cache, node));
+    let value = cache.peek();
+
+    if (isSafeString(value)) {
+      let bounds = vm.stack().insertHTMLBefore(null, value);
```

Comment:
Ember doesn't actually implement that optimization for simple helpers yet, but it will be pretty simple to do that. Here are some examples of returning const references: https://github.com/emberjs/ember.js/pull/13332/files

* In `{{component "foo" ...}}` we can notice that foo is const (`isConst(args.positional.at(0)`), therefore this component helper cannot change to a different component type, which allows us to optimize out some updating stuff
* In `{{if @cond "foo" bar}}`, this has to be dynamic, because `@cond` and `bar` could change. However, if we are able to know that `@cond` is hard-coded to `true` from the invocation template, then we know we will not ever go into the `bar` branch, so the whole thing is actually `isConst` (always the static string "foo")